src/main/kotlin/com/pubnub/internal/v2/subscription/SubscriptionImpl.kt

```diff
@@ -0,0 +1,72 @@
+package com.pubnub.internal.v2.subscription
+
+import com.pubnub.api.PubNub
+import com.pubnub.api.models.consumer.pubsub.PNEvent
+import com.pubnub.api.v2.callbacks.EventEmitter
+import com.pubnub.api.v2.callbacks.EventListener
+import com.pubnub.api.v2.subscriptions.Subscription
+import com.pubnub.api.v2.subscriptions.SubscriptionCursor
+import com.pubnub.api.v2.subscriptions.SubscriptionOptions
+import com.pubnub.api.v2.subscriptions.SubscriptionSet
+import com.pubnub.internal.v2.callbacks.EventEmitterImpl
+import com.pubnub.internal.v2.entities.ChannelGroupName
+import com.pubnub.internal.v2.entities.ChannelName
+import org.jetbrains.annotations.TestOnly
+
+internal class SubscriptionImpl(
+    internal val pubnub: PubNub,
+    channels: Set<ChannelName> = emptySet(),
+    channelGroups: Set<ChannelGroupName> = emptySet(),
+    options: SubscriptionOptions = SubscriptionOptions.Default,
+) : Subscription(), EventEmitter {
+
+    internal val channels = channels.toSet()
+    internal val channelGroups = channelGroups.toSet()
+
+    private val filters = options.allOptions.filterIsInstance<Filter>()
+
+    @get:TestOnly
+    internal val eventEmitter = EventEmitterImpl(pubnub, this::accepts)
```

Comment:
in the ADR we agreed that Status events are a property of the PubNub client, and not a specific subscription, so they're on the PN level only.

For example, if you already had a Subscription to let's say "channelA", and you create another one, you wouldn't get a PNStatus, because the new subscribe call is a no-op